runtime/src/main/java/com/dylibso/chicory/runtime/HostImports.java

```diff
@@ -146,6 +163,245 @@ public static final class Builder {
+            }
+
+            public Builder withTable(Limits limits) {
+                return builder.addTable(
+                        new HostTable(
+                                moduleName,
+                                fieldName,
+                                new TableInstance(new Table(ValueType.FuncRef, limits))));
+            }
+
+            public Builder withTable(int min) {
+                return builder.addTable(
+                        new HostTable(
+                                moduleName,
+                                fieldName,
+                                new TableInstance(new Table(ValueType.FuncRef, new Limits(min)))));
+            }
+
+            public Builder withTable(int min, int max) {
+                return builder.addTable(
+                        new HostTable(
+                                moduleName,
+                                fieldName,
+                                new TableInstance(
+                                        new Table(ValueType.FuncRef, new Limits(min, max)))));
+            }
+
+            // TODO: check - this mechanism is limited by the available primitive functional
+            // interfaces
+            public Builder withProcedure(Runnable f) {
+                return builder.addFunction(
+                        new HostFunction(
+                                (Instance instance, Value... args) -> {
+                                    f.run();
+                                    return new Value[] {};
+                                },
+                                moduleName,
+                                fieldName,
+                                List.of(),
+                                List.of()));
+            }
+
+            public Builder withProcedure(Function<Instance, Runnable> consumer) {
+                return builder.addFunction(
+                        new HostFunction(
+                                (Instance instance, Value... args) -> {
+                                    consumer.apply(instance).run();
+                                    return new Value[] {};
+                                },
+                                moduleName,
+                                fieldName,
+                                List.of(),
+                                List.of()));
+            }
+
+            public Builder withSupplier(IntSupplier f) {
+                return builder.addFunction(
+                        new HostFunction(
+                                (Instance instance, Value... args) ->
+                                        new Value[] {Value.i32(f.getAsInt())},
+                                moduleName,
+                                fieldName,
+                                List.of(),
+                                List.of(ValueType.I32)));
+            }
+
+            public Builder withSupplier(Function<Instance, IntSupplier> consumer) {
```

Comment:
The more I look at this approach, the less I'm convinced this is the way to go.

```java
public Builder withSupplier(Function<Instance, IntSupplier> consumer)
```
and
```java
public Builder withSupplier(Function<Instance, LongSupplier> consumer)
```
are going to be not distinguishable after type erasure.

Currently, I think that leaving only:

```java
public Builder withFunction(Function<Instance, Function<Value[], Value[]>> consumer)
```

and generating the binding out of real modules with something like: https://github.com/andreaTP/chicory-bindgen-poc is the best bet.